homeassistant/components/climate/ambiclimate.py

```diff
@@ -0,0 +1,271 @@
+    vol.Required(ATTR_NAME): cv.string,
+    vol.Required(ATTR_VALUE): cv.string,
+})
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the Ambiclimate heater."""
+    import ambiclimate
+    callback_url = '{}{}'.format(hass.config.api.base_url, AUTH_CALLBACK_PATH)
+    oauth = ambiclimate.AmbiclimateOAuth(config.get(CONF_CLIENT_ID),
+                                         config.get(CONF_CLIENT_SECRET),
+                                         callback_url,
+                                         async_get_clientsession(hass))
+
+    store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+    token_info = await store.async_load()
+
+    if not token_info:
+        _LOGGER.info("no token; requesting authorization")
+        hass.http.register_view(AmbiclimateAuthCallbackView(config,
+                                                            async_add_entities,
+                                                            oauth,
+                                                            store))
+
+        def _call_request_config():
+            configurator = hass.components.configurator
+            req = configurator.request_config
+            data = req(DEFAULT_NAME, lambda _: None,
+                       link_name=CONFIGURATOR_LINK_NAME,
+                       link_url=oauth.get_authorize_url(),
+                       description=CONFIGURATOR_DESCRIPTION,
+                       submit_caption=CONFIGURATOR_SUBMIT_CAPTION)
+            hass.data[AMBICLIMATE_DATA] = data
+
+        hass.async_add_job(_call_request_config)
+        return
+
+    if hass.data.get(AMBICLIMATE_DATA):
+        def _call_request_config_done():
+            configurator = hass.components.configurator
+            configurator.request_done(hass.data.get(AMBICLIMATE_DATA))
+            del hass.data[AMBICLIMATE_DATA]
+
+        hass.async_add_job(_call_request_config_done)
+
+    websession = async_get_clientsession(hass)
+    data_connection = ambiclimate.AmbiclimateConnection(oauth,
+                                                        token_info=token_info,
+                                                        websession=websession)
+
+    if await data_connection.refresh_access_token():
+        await store.async_save(data_connection.token_info)
+
+    if not await data_connection.find_devices():
+        return
+
+    devs = []
+    for heater in data_connection.get_devices():
+        await heater.update_device_info()
+        devs.append(Ambiclimate(heater, store))
+
+    async_add_entities(devs, True)
+
+    async def send_comfort_feedback(service):
+        """Send comfort feedback."""
+        device_name = service.data.get(ATTR_NAME)
+        device = data_connection.find_device_by_room_name(device_name)
+        if device:
+            await device.set_comfort_feedback(service.data.get(ATTR_VALUE))
+
+    hass.services.async_register(DOMAIN,
```

Comment:
We don't want platforms registering services anymore -- move this to the component.